### ISSUE TITLE
[API Proposal]: MakeWeakTypeReference as trimming indicator

### ISSUE THREAD

**Issue by hez2010** [dotnet/runtime#74307]
### Background and motivation

While we were working with trimming, we found that we often want to express "code only needed if a type is preserved", but currently we do not have such thing in C#.

For example, in XAML frameworks, we use URLs to locate resources. With source generator, the code of the resource locator may consist with a method with plenty of if statements, for example:

```csharp
// auto-generated by a source generator
object? LoadResource(string path)
{
    if (path == "ms-appx://MyAssembly/My/Fancy/XamlControl1Style") return XamlLoader.Load<XamlControl1Style>();
    if (path == "ms-appx://MyAssembly/Another/XamlControl2Style") return XamlLoader.Load<XamlControl2Style>();
    return null;
}
```

With this source generator approach, we can achieve trimming compatibility. However, this would result in all types referenced in `LoadResource` being preserved even they are not being used actually. While actually we only want `XamlControl1Style` to be preserved if `XamlControl1` is referenced in other pieces of code. 

### API Proposal

```csharp
namespace System.Diagnostics.CodeAnalysis;

public class ControlFlow
{
    public static bool MakeWeakTypeReference<T>() => true;
}
```

### API Usage

```csharp
object? LoadResource(string path)
{
    if (ControlFlow.MakeWeakTypeReference<XamlControl1>())
        if (path == "ms-appx://MyAssembly/My/Fancy/XamlControl1Style")
            return XamlLoader.Load<XamlControl1Style>();
    if (ControlFlow.MakeWeakTypeReference<XamlControl2>())
        if (path == "ms-appx://MyAssembly/Another/XamlControl2Style")
            return XamlLoader.Load<XamlControl2Style>();
    return null;
}
```

It behaves as a no-op which always returns true, but if there are no other type references to `XamlControl1` or `XamlControl2` except `MakeWeakTypeReference`, the `XamlControl1Style` or `XamlControl2Style` can be safely trimmed, respectively.

**Comment by teo-tsirpanis** [comment]
Interesting idea; the name does not sound right to me though.

**Comment by hez2010** [comment]
> Interesting idea; the name does not sound right to me though.

Yes. I did think about the name for a while but didn't conclude a fancy name :)

**Comment by MichalPetryka** [comment]
Maybe `IsTypePresent<T>`?

**Comment by teo-tsirpanis** [comment]
That's better, and with that name it fits more on `Type`.

**Comment by eerhardt** [comment]
See the following code, which we use in System.Text.Json to do this exact scenario:

https://github.com/dotnet/runtime/blob/e71a9583b4d6c9bd97edd87cda7f98f232f63530/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactoryHelpers.cs#L120-L125

**Comment by jkoritzinsky** [comment]
@eerhardt that's really clever and I think a good example for where an API like this could be useful. I think being able to have a well-known API that would enable this to work while using generics or strongly-typed `System.Type` objects instead of having to round-trip through type name parsing would be quite useful and likely more maintainable than the existing mechanism you shared.

**Comment by MichalPetryka** [comment]
> @eerhardt that's really clever and I think a good example for where an API like this could be useful. I think being able to have a well-known API that would enable this to work while using generics or strongly-typed `System.Type` objects instead of having to round-trip through type name parsing would be quite useful and likely more maintainable than the existing mechanism you shared.

Additionally, it'd be more AOT safe.

**Comment by teo-tsirpanis** [comment]
```csharp
namespace System;

public abstract partial class Type
{
    public static Type? GetTypeIfExists<T>() => typeof(T); // would be replaced with null by the trimmer if the type does not exist.
}
```

**Comment by kekekeks** [comment]
I believe that just getting the `Type` instance conditionally isn't enough for XAML styles. Not sure how it works in the case of the code generator mentioned in the issue, but with Avalonia XAML Styles will usually reference the properties of the type itself from MSIL.

e. g. we have something like this (equivalent C# translation, the actual XAML compiler generates msil): 

```cs
Application.Resources[typeof(Button)] = new ControlTheme(typeof(Button))
{
  new Setter(Button.TemplateProperty, new FuncControlTemplate<Button>(() => new ContentPresenter
  {
      [ContentPresenter.ContentProperty] = new TemplateBinding(Button.ContentProperty),
      [ContentPresenter.PaddingProperty] = new TemplateBinding(Button.PaddingProperty)
  }));,
  new Setter(Button.PaddingProperty, new Padding(4)),
  ...
}
```

During linking if Button class isn't used, it's needed to trim the entire code that constructs the Button's theme, because everything is strongly typed and liker we'll see references from MSIL.

**Comment by MichalPetryka** [comment]
Working on `Type` would also be less AOT and linker friendly since for example the AOT wouldn't be able to MakeGenericType with types provided that way.

**Comment by hez2010** [comment]
> During linking if Button class isn't used, it's needed to trim the entire code that constructs the Button's theme, because everything is strongly typed and liker we'll see references from MSIL, so the code would have to look like this:
> 
> ```cs
> if(Type.Exists<Button>())
> {
>   Application.Resources[typeof(Button)] = new ControlTheme(typeof(Button))
>   {
>     new Setter(Button.TemplateProperty, new FuncControlTemplate<Button>(() => new ContentPresenter
>     {
>         [ContentPresenter.ContentProperty] = new TemplateBinding(Button.ContentProperty),
>         [ContentPresenter.PaddingProperty] = new TemplateBinding(Button.PaddingProperty)
>     }));,
>     new Setter(Button.PaddingProperty, new Padding(4)),
>     ...
>   };
> }
> ```
> 
> with `if` body trimmed away if `Button` class isn't referenced elsewhere.

I think `if (Type.GetTypeIfExists<Button>() != null)` is equivalent to `if (Type.Exists<Button>())`.

**Comment by vitek-karas** [comment]
This is another manifestation of a problem which we've ran into couple of times already but so far didn't really tackle. Basically source generators have a hard time attaching the generated code to a "target" type. The usual recommendation is to make the target type partial and thus let the source generator adds stuff to it. But that's not always possible, for example in the above case of `Button` class, there's no good way to attach anything to it from the app's code.

So people solve this by introducing some types of registration mechanism. System.Text.Json uses the "context" for example. If such registration is global, it naturally leads to rooting everything in the registry regardless if it's needed by the app or not.

There was a similar discussion on the topic in this issue https://github.com/dotnet/runtime/issues/50333. The original take is somewhat different, but it discusses basically the same problem space.

I do like the approach proposed in this issue which uses "code", instead of declarative attributes or similar solutions. But as proposed this goes against one of the main design principles of trimming:
> Trimming an app should not change any observable behavior of the code

This proposal would explicitly add a public API which changes behavior when trimmed. As used above it's OK, but one could easily use it to change other types of behavior and thus introduce unpredictability into the development process (debug build should behave the same as published app).

Trimmer already has a feature which goes into this area:
```C#
void Test(object o)
{
    if (o is SpecialType)
        Console.WriteLine("Found something special");
}
```

If this is the only reference to `SpecialType` in the app, trimmer will actually remove `SpecialType` and rewrite the condition to `if (false)`. But it keeps the if branch around, along with all of its dependencies. (see [tests](https://github.com/dotnet/linker/blob/main/test/Mono.Linker.Tests.Cases/Advanced/TypeCheckRemoval.cs)).

If we extended this to actually remove the if branch it could be used to provide a solution the problem stated in this issue. The advantage of such solution would be that there's no observable difference when the app is trimmed. If the `SpecialType` is not used anywhere in the app, the if branch would never execute anyway (trimming or not), so it's effectively dead code and can be removed.

It would require a slight design change in the above samples, but it feels possible. I can also imagine supporting the `Type` version of this:
```C#
void Test(Type type)
{
    if (type == typeof(SpecialType))
        Console.WriteLine(...);
}
```

Although this might need more thinking if it's truly correct.

Note that the "instance" version of this check is actually slightly better: If the `SpecialType` is used elsewhere, but it's never instantiated the condition can still be optimized away (as it will always be false). This is not true for the "type" variant.

**Comment by MichalStrehovsky** [comment]
I agree with Vitek that is this proposing an API addition that goes against how we want trimming to work - the trimmed app should behave the same as untrimmed app by default. Whenever there's a behavior difference, there should be a warning.

We would need to immediately tag this API as [`RequiresUnreferencedCode`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.requiresunreferencedcodeattribute?view=net-6.0) so that a warning is triggered. Suppressing trimming warnings is a can of worms and we very much recommend against suppressing any warnings. We were not able to get warning suppressions right in this very repo with all the expertise we have. I have zero confidence in third parties to get their suppressions right.

Besides the philosophical problem above, the problem I have with an API like this is that it relies on branch removal. Branch removal does happen both in IL Linker and Native AOT. It is also done with different algorithms, and I would like to see the NativeAOT one to be replaced with whatever constprop RyuJIT can do, which is yet another algorithm.

We didn't document branch removal rules on purpose. We only test them on the runtime repo. We had to [adjust code](https://github.com/dotnet/runtime/pull/40539) in the runtime repo to fit the envelope of the optimization in the past. This would make relying on the shape of the optimization part of a public API surface (the API by itself has no legitimate use outside of this pattern). There would be no feedback if constprop is not able to kick in and doesn't eliminate a branch that the user expects. This is hard to debug for end users.

The other thing is that "a type is preserved" is a very murky concept:

```csharp
class Program
{
    static void Do(SomeClass x) { }

    static void Main()
    {
         // Is SomeClass preserved? Trimming at IL Level would need to preserve it
        // because it’s referenced from the method signature. But it doesn’t exist for NativeAOT purposes
        // except in the PDB.
        Do(null);

        // Is SomeClass preserved now?
        // Assume SomeSimpleStaticMethod was simple enough that IL Linker inlined it.
        SomeClass.SomeSimpleStaticMethod();

        // How about now?
        // SomeClass would be reflection visible now when IL trimming. NativeAOT still has leeway.
        SomeClass.OtherSimpleStaticMethod();
    }
}
```

This API is basically trying to query at runtime for the result of a compile time optimization that is supposed to be transparent.

---------

I would really like to think more in directions that can have the same behavior with/without trimming.

For the XAML example, could the Button class in its constructor do a callout that registers the styles? I assume we don't need styles for a button if no button was created yet.

**Comment by kekekeks** [comment]
The problem with XAML styles is that they are completely separate from the control itself. Control does not and should not know anything about its styles, they are essentially user-defined and can be switched to a separate set of styles at runtime during the theme switch. 
Just like in WPF, in Avalonia styles are defined in a separate assembly (in case of Avalonia they come in separate nuget packages) and can be replaced for the entire app or a particular control subtree with another set of styles that come from a 3rd party package.

**Comment by kekekeks** [comment]
Another use case:

My [COM interop bindings](https://github.com/kekekeks/MicroCom) generate COM interfaces as normal .NET interfaces.

e. g. 
```
[uuid(ED7F3339-8EB4-4912-9D6F-EC4CDD094164})]
interface IFoo : IUnknown
{
	void Bar();
}
```

becomes
```cs
// IUnknown is defined as a marker interface on top of IDisposable
interface IFoo : IUnknown
{
    void Bar();
}
```



Interop is handled by separate generated classes `__MicroComIFooProxy` (implements the managed interface and forwards calls to native function pointers) and `__MicroComIFooVTable` (builds a CCW vtable).

Both are registered by the module initializer. I'd like to be able to trim unneeded interop types when the interface isn't used.

**Comment by jkotas** [comment]
> Interop is handled by separate generated classes __MicroComIFooProxy (implements the managed interface and forwards calls to native function pointers) and __MicroComIFooVTable (builds a CCW vtable).

CsWinRT solved this problem by linking the interop helper types via custom attributes: https://github.com/microsoft/CsWinRT/pull/1224

**Comment by kekekeks** [comment]
I believe such custom attributes require reflection access. MicroCom aims to work with IlcDisableReflection=True

**Comment by teo-tsirpanis** [comment]
No, the attributes would be accessed at the time of the source generation, not at runtime.

**Comment by jkotas** [comment]
> type.GetCustomAttribute looks like run-time reflection

Yes, custom attributes are accessed via reflection. The key advantage is that the helper types associated via custom attribute can be statically analyzed for correctness and the solution does not have concerns shared by @vitek-karas and @MichalStrehovsky above.

>  var helper = $"ABI.{fullTypeName}";
> return type.Assembly.GetType(helper) ?? Type.GetType(helper);

This is temporary fallback path that should go away eventually.

**Comment by MichalStrehovsky** [comment]
> I believe such custom attributes require reflection access. MicroCom aims to work with IlcDisableReflection=True

Working with IlcDisableReflection is a noble goal but I don't think the base class libraries will ever be in a position where they work with IlcDisableReflection=true. They likely won't even work in the more usable future version of it: https://github.com/dotnet/runtime/issues/67193#issuecomment-1214229147. Trimmable reflection is not a problem and would be a good solution for this.

IlcDisableReflection is a tech demo that was introduced before we had things like IlcTrimMetadata=true (which is now the default). With metadata trimming, I don't believe you'll see meaningful size differences with/without reflection for apps that are bigger than a console hello world. IlcDisableReflection produces meaningful savings for hello-world-sized apps because it allows removing the whole reflection stack, but that's a fixed ~1MB cost that is not worth it for most. We'll likely do work that will indirectly make this cost smaller over time. The reflection stack in NativeAOT is a bit overengineered.

If you still see a _meaningful_ difference for Avalonia-sized apps, please add `<ItemGroup><IlcArg Include="--make-repro-path:c:\some\path\where\a\zip\can\be\created" /></ItemGroup>` to the project and share the resulting ZIP with me (in a new issue) - it's likely a bug - IlcDisableReflection has been useful in finding those in the past.

**Comment by vitek-karas** [comment]
> We didn't document branch removal rules on purpose. We only test them on the runtime repo. We had to https://github.com/dotnet/runtime/pull/40539 in the runtime repo to fit the envelope of the optimization in the past. This would make relying on the shape of the optimization part of a public API surface (the API by itself has no legitimate use outside of this pattern). There would be no feedback if constprop is not able to kick in and doesn't eliminate a branch that the user expects. This is hard to debug for end users.

I view this as two separate things:
* Correctness/consistency - the app behaves the same with/without trimming. The idea of branch removal doesn't violate this in any way. (unlike the proposed API which does)
* Size optimization (trimming) - currently we don't describe nor make any promises with regard to trimming specific pieces of the app. The tools (trimmer, NativeAOT, ...) may decide to keep code/metadata around without any specific documented reason. This leads to us not making any size guarantees. The branch removal proposed above doesn't work today, but if it enabled it, it would simple improve size for some apps, but should have no other impact.

That said, it would be nice to have some agreed upon way especially for libraries to verify that they can be successfully trimmed to a certain degree - and to enable detection or regressions and such. But that's sort of outside the scope of this issue.
